### PR TITLE
Adds StoreInjector and StoreContainer

### LINKED ISSUE #141: Add `StoreInjector` to enable binding a store to a widget Container`
**Enhancement**

At the moment to bind the store to a Container in a widget tree, consumers need to create a custom Injector:

```ts
class StoreInjector extends Injector {
	constructor(payload: Store) {
		super(payload);
		payload.on('invalidate', () => {
			this.emit({ type: 'invalidate' });
		});
	}
}
```

It would be helpful if this was provided by default, either in `dojo/stores` or possibly in `dojo/interop`.

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

**Description:**

Adds a `StoreInjector` and `StoreContainer` that wire the `store` invalidation event to the injector invalidation event. Additionally when creating a `Container` consumers can choose specific parts of a stores state that they are interested in and only changes to these paths should cause an invalidation.

Resolves #141

### CODE REVIEW

File: src/StoreInjector.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
+import { w } from '@dojo/widget-core/d';
+import { handleDecorator } from '@dojo/widget-core/decorators/handleDecorator';
+import { beforeProperties } from '@dojo/widget-core/decorators/beforeProperties';
+import { alwaysRender } from '@dojo/widget-core/decorators/alwaysRender';
+import { RegistryLabel, Constructor, DNode } from '@dojo/widget-core/interfaces';
+import { Store } from './Store';
+import { Injector } from '@dojo/widget-core/Injector';
+
+const registeredInjectorsMap: WeakMap<WidgetBase, StoreInjector[]> = new WeakMap();
```
Comment (tomdye) [inline src/StoreInjector.ts:11]: Do you think this would be better as `WeakMap<WidgetBase, Set<StoreInjector>>` ?

Comment (agubler) [reply src/StoreInjector.ts:11]: Think it's worthwhile introducing `Set` (plus the shim), what benefit does the `Set` get us?

Comment (tomdye) [reply src/StoreInjector.ts:11]: Ok, fair enough. nothing major, was just a thought.

**Review by tomdye — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -350,6 +376,50 @@ store.on('invalidate', () => {
 });
 ```
 
+### Connecting Store Updates To Widgets
+
+Store data can be connected to widgets within your application using the [Containers & Injectors Pattern](https://github.com/dojo/widget-core#containers--injectors) supported by `@dojo/widget-core`. `@dojo/stores` provides a specialized injector that invalidates store containers on two conditions:
+
+1. The recommended approach is to register `paths` on container creation to ensure invalidation will only occur when state you are interested in changes.
```
Comment (dylans) [inline README.md:383]: "will only occur when state you are interested in changes" ?

File: README.md [hunk 3]
```diff
@@ -350,6 +376,50 @@ store.on('invalidate', () => {
...
+2. A catch all when no `paths` are defined for the container, it will invalidate when any data changes in the store.
```
Comment (dylans) [inline README.md:384]: Maybe change this sentence to be a bit more consistent with the previous one as a second condition?

File: README.md [hunk 4]
```diff
@@ -350,6 +376,50 @@ store.on('invalidate', () => {
...
+
+// Catch all, will invalidate if _any_ state changes in the store even if the container is not interested in the changes
+const Container = Container(WidgetBase, 'state', { getProperties(store: Store<State>) {
+	return {
+		foo: store.get(store.path('foo'))
+	}
+}});
+```
+
+To provide a typed container for the stores state, use the `createStoreContainer` function from `@dojo/stores/StoreInjector` passing in the state interface as a generic and export this to be available throughout your application.
```
Comment (dylans) [inline README.md]: stores state -> store's state

Comment (dylans) [inline README.md]: export this -> export _______ (something to make sure its clear what this is)

**Review by dylans — CHANGES_REQUESTED** [review]
a few small grammar nits/suggestions

File: README.md (hunk 2 above)
Comment (agubler) [reply README.md:383]: Perhaps should say when state changes for the paths change?

File: README.md [hunk 5]
```diff
@@ -410,7 +410,7 @@ const Container = Container(WidgetBase, 'state', { getProperties(store: Store<St
 }});
 ```
 
-To provide a typed container for the stores state, use the `createStoreContainer` function from `@dojo/stores/StoreInjector` passing in the state interface as a generic and export this to be available throughout your application.
+To provide a typed container for the store's state, use the `createStoreContainer` function from `@dojo/stores/StoreInjector` passing in the state interface as a generic and then export returned `Container` to be available throughout your application.
```
Comment (dylans) [inline README.md]: export the returned...

**Review by dylans — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (agubler) [reply README.md]: doh

**Review by dylans — APPROVED** [review]

**Pull request merged**